### PR TITLE
All HitBox and CollisionBox Changes made for LineOfSightPlace

### PR DESCRIPTION
**Pull request by Axionize**
As requested by Cotander, I've made a separate PR for all changes made that affect the simulation engine (Hitbox and CollisionBox changes, mostly)

### CODE REVIEW

**Comment by Axionize** [comment]
Currently out of date, need to backport some changes made for LOSP.

**Comment by Axionize** [comment]
This branch is now up to date with all simulation-related (hitbox and collisionbox) bugfixes and changes made while writing LOSP. Feel free to review and merge

File: src/main/java/ac/grim/grimac/utils/collisions/CollisionData.java [hunk 1]
```diff
@@ -830,65 +792,29 @@ public enum CollisionData {
...
-                return new SimpleCollisionBox(0.5 - f, 0.2, 0.0, 0.5 + f, 0.8, f * 2.0, false);
-            case DOWN:
-                return new SimpleCollisionBox(0.25, 0.4, 0.25, 0.75, 1.0, 0.75, false);
-            case UP:
-                return new SimpleCollisionBox(0.25, 0.0, 0.25, 0.75, 0.6, 0.75, false);
-        }
-
-        return NoCollisionBox.INSTANCE;
-
-    }, StateTypes.LEVER),
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/utils/collisions/CollisionData.java]: These should be moved to `NO_COLLISION`

File: src/main/java/ac/grim/grimac/utils/collisions/CollisionData.java [hunk 2]
```diff
@@ -921,24 +847,11 @@ public enum CollisionData {
...
-            default:
-                return new HexCollisionBox(6.0D, 0.0D, 6.0D, 16.0D, 10.0D, 10.0D);
-        }
+        return NoCollisionBox.INSTANCE;
     }, StateTypes.ATTACHED_MELON_STEM, StateTypes.ATTACHED_PUMPKIN_STEM),
 
     PUMPKIN_STEM((player, version, data, x, y, z) -> {
-        return new HexCollisionBox(7, 0, 7, 9, 2 * (data.getAge() + 1), 9);
+        return NoCollisionBox.INSTANCE;
     }, StateTypes.PUMPKIN_STEM, StateTypes.MELON_STEM),
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/utils/collisions/CollisionData.java]: Same as above

**Review by ManInMyVan — CHANGES_REQUESTED** [review]

File: src/main/java/ac/grim/grimac/utils/collisions/blocks/connecting/DynamicCollisionWall.java [hunk 3]
```diff
@@ -15,12 +15,13 @@
...
 import com.github.retrooper.packetevents.protocol.world.states.type.StateType;
 
-public class DynamicWall extends DynamicConnecting implements CollisionFactory {
+public class DynamicCollisionWall extends DynamicConnecting implements CollisionFactory {
     public static final CollisionBox[] BOXES = makeShapes(4.0F, 3.0F, 16.0F, 0.0F, 16.0F, false);
     // https://bugs.mojang.com/browse/MC-9565
     // https://bugs.mojang.com/browse/MC-94016
     private static final CollisionBox[] COLLISION_BOXES = makeShapes(4.0F, 3.0F, 24.0F, 0.0F, 24.0F, false);
 
+    // Deprecated in favor of DynamicHitboxWall
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/utils/collisions/blocks/connecting/DynamicCollisionWall.java]: Remove this method if we aren't using it anywhere, or mark it `@Deprecated`
```suggestion
    /**
     * @deprecated use DynamicHitboxWall
     */
    @Deprecated
```

File: src/main/java/ac/grim/grimac/utils/collisions/HitboxData.java [hunk 4]
```diff
@@ -264,24 +252,284 @@ public enum HitboxData {
...
+        }
+    }, StateTypes.BREWING_STAND),
+
+    SMALL_FLOWER((player, item, version, data, x, y, z) ->  player.getClientVersion().isOlderThan(ClientVersion.V_1_13)
+            ? new SimpleCollisionBox(0.3125D, 0.0D, 0.3125D, 0.6875D, 0.625D, 0.6875D)
+            : new OffsetCollisionBox(data.getType(), 0.3125D, 0.0D, 0.3125D, 0.6875D, 0.625D, 0.6875D),
+            BlockTags.SMALL_FLOWERS.getStates().toArray(new StateType[0])),
+
+    TALL_FLOWERS((player, item, version, data, x, y, z) -> new SimpleCollisionBox(0, 0, 0, 1, 1, 1, true),
+            BlockTags.TALL_FLOWERS.getStates().toArray(new StateType[0])),
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/utils/collisions/HitboxData.java]:
```suggestion
    TALL_FLOWERS(new SimpleCollisionBox(0, 0, 0, 1, 1, 1, true), BlockTags.TALL_FLOWERS.getStates().toArray(new StateType[0])),
```

File: src/main/java/ac/grim/grimac/utils/collisions/CollisionData.java [hunk 5]
```diff
@@ -281,43 +261,36 @@ public enum CollisionData {
...
-    SWEET_BERRY((player, version, data, x, y, z) -> {
-        if (data.getAge() == 0) {
-            return new HexCollisionBox(3.0D, 0.0D, 3.0D, 13.0D, 8.0D, 13.0D);
-        } else if (data.getAge() < 3) {
-            return new HexCollisionBox(1.0D, 0.0D, 1.0D, 15.0D, 16.0D, 15.0D);
-        }
-        return new SimpleCollisionBox(0, 0, 0, 1, 1, 1, true);
-    }, StateTypes.SWEET_BERRY_BUSH),
+    RAILS(NoCollisionBox.INSTANCE,
+            BlockTags.RAILS.getStates().toArray(new StateType[0])),
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/utils/collisions/CollisionData.java]:
```suggestion
    RAILS(NoCollisionBox.INSTANCE, BlockTags.RAILS.getStates().toArray(new StateType[0])),
```

File: src/main/java/ac/grim/grimac/utils/collisions/CollisionData.java [hunk 6]
```diff
@@ -210,9 +193,6 @@ public enum CollisionData {
         }
     }, StateTypes.PIGLIN_WALL_HEAD),
 
-    BANNER(new HexCollisionBox(4.0D, 0.0D, 4.0D, 12.0D, 16.0D, 12.0D),
-            BlockTags.BANNERS.getStates().toArray(new StateType[0])),
-
     CORAL_FAN((player, version, data, x, y, z) -> {
         return new HexCollisionBox(2.0D, 0.0D, 2.0D, 14.0D, 4.0D, 14.0D);
     }, BlockTags.CORALS.getStates().toArray(new StateType[0])),
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/utils/collisions/CollisionData.java]:
```suggestion
    CORAL_FAN(new HexCollisionBox(2.0D, 0.0D, 2.0D, 14.0D, 4.0D, 14.0D), BlockTags.CORALS.getStates().toArray(new StateType[0])),
```

File: src/main/java/ac/grim/grimac/utils/collisions/CollisionData.java [hunk 7]
```diff
@@ -281,43 +261,36 @@ public enum CollisionData {
...
+    RAILS(NoCollisionBox.INSTANCE,
+            BlockTags.RAILS.getStates().toArray(new StateType[0])),
 
-    SAPLING(new HexCollisionBox(2.0D, 0.0D, 2.0D, 14.0D, 12.0D, 14.0D),
-            BlockTags.SAPLINGS.getStates().toArray(new StateType[0])),
-
-    ROOTS(new HexCollisionBox(2.0D, 0.0D, 2.0D, 14.0D, 13.0D, 14.0D),
-            StateTypes.WARPED_ROOTS, StateTypes.CRIMSON_ROOTS),
+    BANNER(NoCollisionBox.INSTANCE,
+            BlockTags.BANNERS.getStates().toArray(new StateType[0])),
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/utils/collisions/CollisionData.java]:
```suggestion
    BANNER(NoCollisionBox.INSTANCE, BlockTags.BANNERS.getStates().toArray(new StateType[0])),
```

File: src/main/java/ac/grim/grimac/utils/collisions/CollisionData.java [hunk 8]
```diff
@@ -281,43 +261,36 @@ public enum CollisionData {
...
 
-    FLOWER(new HexCollisionBox(5.0D, 0.0D, 5.0D, 11.0D, 10.0D, 11.0D),
+    SMALL_FLOWER(NoCollisionBox.INSTANCE,
             BlockTags.SMALL_FLOWERS.getStates().toArray(new StateType[0])),
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/utils/collisions/CollisionData.java]:
```suggestion
    SMALL_FLOWER(NoCollisionBox.INSTANCE, BlockTags.SMALL_FLOWERS.getStates().toArray(new StateType[0])),
```

File: src/main/java/ac/grim/grimac/utils/collisions/CollisionData.java [hunk 9]
```diff
@@ -281,43 +261,36 @@ public enum CollisionData {
...
-
-    GRASS_FERN(new HexCollisionBox(2.0D, 0.0D, 2.0D, 14.0D, 13.0D, 14.0D),
-            StateTypes.GRASS, StateTypes.FERN),
-
-    TALL_GRASS(new SimpleCollisionBox(0, 0, 0, 1, 1, 1, true), StateTypes.TALL_GRASS),
-
-    SEA_GRASS(new HexCollisionBox(2.0D, 0.0D, 2.0D, 14.0D, 12.0D, 14.0D),
-            StateTypes.SEAGRASS),
+    TALL_FLOWER(NoCollisionBox.INSTANCE,
+            BlockTags.TALL_FLOWERS.getStates().toArray(new StateType[0])),
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/utils/collisions/CollisionData.java]:
```suggestion
    TALL_FLOWER(NoCollisionBox.INSTANCE, BlockTags.TALL_FLOWERS.getStates().toArray(new StateType[0])),
```

File: src/main/java/ac/grim/grimac/utils/collisions/CollisionData.java [hunk 10]
```diff
@@ -281,43 +261,36 @@ public enum CollisionData {
...
 
-    CAVE_VINES(new HexCollisionBox(1.0D, 0.0D, 1.0D, 15.0D, 16.0D, 15.0D), StateTypes.CAVE_VINES, StateTypes.CAVE_VINES_PLANT),
+    SAPLING(NoCollisionBox.INSTANCE,
+            BlockTags.SAPLINGS.getStates().toArray(new StateType[0])),
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/utils/collisions/CollisionData.java]:
```suggestion
    SAPLING(NoCollisionBox.INSTANCE, BlockTags.SAPLINGS.getStates().toArray(new StateType[0])),
```

File: src/main/java/ac/grim/grimac/utils/collisions/CollisionData.java [hunk 11]
```diff
@@ -281,43 +261,36 @@ public enum CollisionData {
...
 
-    TWISTING_VINES_BLOCK(new HexCollisionBox(4.0D, 0.0D, 4.0D, 12.0D, 15.0D, 12.0D), StateTypes.TWISTING_VINES, StateTypes.WEEPING_VINES),
+    BUTTON((player, version, data, x, y, z) -> {
+        return NoCollisionBox.INSTANCE;
+    }, BlockTags.BUTTONS.getStates().toArray(new StateType[0])),
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/utils/collisions/CollisionData.java]:
```suggestion
    BUTTON(NoCollisionBox.INSTANCE, BlockTags.BUTTONS.getStates().toArray(new StateType[0])),
```

File: src/main/java/ac/grim/grimac/utils/collisions/CollisionData.java [hunk 12]
```diff
@@ -281,43 +261,36 @@ public enum CollisionData {
...
 
-    TWISTING_VINES(new HexCollisionBox(4.0D, 0.0D, 4.0D, 12.0D, 16.0D, 12.0D), StateTypes.TWISTING_VINES_PLANT, StateTypes.WEEPING_VINES_PLANT),
+    // All states that unconditionally have no collision and are not in a group where ever member also has no collision.
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/utils/collisions/CollisionData.java]: "ever member" doesn't make sense, did you mean "every member"?

File: src/main/java/ac/grim/grimac/utils/collisions/HitboxData.java [hunk 13]
```diff
@@ -264,24 +252,284 @@ public enum HitboxData {
...
+            case NORTH:
+                return new HexCollisionBox(0.0, 0.0, 14.0, 16.0, 12.5, 16.0);
+            case EAST:
+                return new HexCollisionBox(0.0, 0.0, 0.0, 2.0, 12.5, 16.0);
+            case WEST:
+                return new HexCollisionBox(14.0, 0.0, 0.0, 16.0, 12.5, 16.0);
+            case SOUTH:
+                return new HexCollisionBox(0.0, 0.0, 0.0, 16.0, 12.5, 2.0);
+            default:
+                throw new RuntimeException("Impossible Banner Facing State; Something very wrong is going on");
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/utils/collisions/HitboxData.java]: imo `RuntimeException` should be replaced with something else like `IllegalStateException` or `IllegalArgumentException`

File: src/main/java/ac/grim/grimac/utils/collisions/HitboxData.java [hunk 14]
```diff
@@ -264,24 +252,284 @@ public enum HitboxData {
...
+    COCOA_BEANS((player, item, version, data, x, y, z) -> {
+        return CollisionData.getCocoa(version, data.getAge(), data.getFacing());
+    }, StateTypes.COCOA),
+
+
+    REDSTONE_WIRE((player, item, version, data, x, y, z) ->
+            // Easier to just use no collision box
+            // Redstone wire is very complex with its collision shapes and has many de-syncs
+            NoCollisionBox.INSTANCE,
+            StateTypes.REDSTONE_WIRE),
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/utils/collisions/HitboxData.java]:
```suggestion
    // Easier to just use no collision box
    // Redstone wire is very complex with its collision shapes and has many de-syncs
    REDSTONE_WIRE(NoCollisionBox.INSTANCE, StateTypes.REDSTONE_WIRE),
```

File: src/main/java/ac/grim/grimac/utils/collisions/HitboxData.java [hunk 15]
```diff
@@ -264,24 +252,284 @@ public enum HitboxData {
...
+    SUGARCANE(new HexCollisionBox(2.0D, 0.0D, 2.0D, 14.0D, 16.0D, 14.0D), StateTypes.SUGAR_CANE),
+
+    NETHER_SPROUTS(new HexCollisionBox(2.0D, 0.0D, 2.0D, 14.0D, 3.0D, 14.0D), StateTypes.NETHER_SPROUTS),
+
+    HANGING_ROOTS(new HexCollisionBox(2.0D, 10.0D, 2.0D, 14.0D, 16.0D, 14.0D), StateTypes.HANGING_ROOTS),
+
+    GRASS_FERN((player, item, version, data, x ,y, z) -> {
+        if (version.isOlderThan(ClientVersion.V_1_13)) {
+            float var1 = 0.4F;
+            return new SimpleCollisionBox(0.5F - var1, 0.0F, 0.5F - var1, 0.5F + var1, 0.8F, 0.5F + var1);
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/utils/collisions/HitboxData.java]:
```suggestion
            return new SimpleCollisionBox(0.1F, 0.0F, 0.1F, 0.9F, 0.8F, 0.9F);
```

File: src/main/java/ac/grim/grimac/utils/collisions/HitboxData.java [hunk 16]
```diff
@@ -264,24 +252,284 @@ public enum HitboxData {
...
+
+        return box;
+
+    }, StateTypes.BIG_DRIPLEAF),
+
+    MANGROVE_PROPAGULE(((player, item, version, data, x, y, z) -> {
+        if (data.isHanging()) {
+            return new HexOffsetCollisionBox(data.getType(), 7.0, 0.0, 7.0, 9.0, 16.0, 9.0);
+        } else {
+            return new HexOffsetCollisionBox(data.getType(),7.0, getPropaguleMinHeight(data.getAge()), 7.0, 9.0, 16.0, 9.0);
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/utils/collisions/HitboxData.java]:
```suggestion
            return new HexOffsetCollisionBox(data.getType(), 7.0, getPropaguleMinHeight(data.getAge()), 7.0, 9.0, 16.0, 9.0);
```

File: src/main/java/ac/grim/grimac/utils/collisions/HitboxData.java [hunk 17]
```diff
@@ -329,6 +577,22 @@ public static CollisionBox getBlockHitbox(GrimPlayer player, StateType heldItem,
...
+        switch (age) {
+            case 0:
+            case 1:
+            case 2:
+                return 13 - age * 3;
+            case 3:
+            case 4:
+                return (4 - age) * 3;
+        }
+        throw new RuntimeException("Impossible State");
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/utils/collisions/HitboxData.java]: imo `RuntimeException` should be replaced with something else like `IllegalStateException` or `IllegalArgumentException`

**Review by ManInMyVan — CHANGES_REQUESTED** [review]

File: src/main/java/ac/grim/grimac/utils/math/GrimMath.java [hunk 18]
```diff
@@ -105,4 +105,10 @@ public static boolean inRange(double value, double min, double max) {
     public static boolean isNearlySame(double a, double b, double epoch) {
         return Math.abs(a-b) < epoch;
     }
+
+    public static long hashCode(double x, int y, double z) {
+        long l = (long)(x * 3129871) ^ (long)z * 116129781L ^ (long)y;
+        l = l * l * 42317861L + l * 11L;
+        return l >> 16;
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/utils/math/GrimMath.java:112]: This should return the same result (test this)
You should also rename `l`
```suggestion
        long l = x * 3129871 ^ z * 116129781L ^ y;
        return l * l * 42317861 + l * 11 >> 16;
```

File: src/main/java/ac/grim/grimac/utils/collisions/datatypes/SimpleCollisionBox.java [hunk 19]
```diff
@@ -20,6 +20,7 @@ public class SimpleCollisionBox implements CollisionBox {
     public double minX, minY, minZ, maxX, maxY, maxZ;
     private boolean isFullBlock = false;
 
+    // If your min/max values are > 1 you should probably check out HexCollisionBox
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/utils/collisions/datatypes/SimpleCollisionBox.java]: shouldn't this comment be on the constructor with min/max parameters?

also use javadoc, might need to import `HexCollisionBox` to use `@link`
```suggestion
    /**
     * If your min/max values are > 1 you should probably check out {@link HexCollisionBox}
     */
```

**Review by ManInMyVan — CHANGES_REQUESTED** [review]

File: src/main/java/ac/grim/grimac/utils/math/GrimMath.java (hunk 18 above)
Comment (Axionize) [reply src/main/java/ac/grim/grimac/utils/math/GrimMath.java:112]: This will also not work. You need to cast to long for ^

**Comment by Axionize** [comment]
Thank you Sam and ManInMyVan. I apologize for the massive commit but the lack of foresight into the size of the changes required to make this work and a stacked diff system kind of forced my hand.

**Pull request merged**